File: src/apps/kernel/arch/gem5/Platform.cc
/*
 * Copyright (C) 2015, Nils Asmussen <nils@os.inf.tu-dresden.de>
 * Economic rights: Technische Universitaet Dresden (Germany)
 *
 * This file is part of M3 (Microkernel-based SysteM for Heterogeneous Manycores).
 *
 * M3 is free software: you can redistribute it and/or modify
 * it under the terms of the GNU General Public License version 2 as
 * published by the Free Software Foundation.
 *
 * M3 is distributed in the hope that it will be useful, but
 * WITHOUT ANY WARRANTY; without even the implied warranty of
 * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
 * General Public License version 2 for more details.
 */

#include <base/Init.h>

#include "mem/MainMemory.h"
#include "mem/MemoryModule.h"
#include "pes/VPE.h"
#include "DTU.h"
#include "Platform.h"

namespace kernel {

INIT_PRIO_USER(2) Platform::KEnv Platform::_kenv;

// note that we currently assume here, that compute PEs and memory PEs are not mixed
static peid_t last_pe_id;

Platform::KEnv::KEnv() {
    // read kernel env
    peid_t pe = m3::DTU::gaddr_to_pe(m3::env()->kenv);
    goff_t addr = m3::DTU::gaddr_to_virt(m3::env()->kenv);
    DTU::get().read_mem(VPEDesc(pe, VPE::INVALID_ID), addr, this, sizeof(*this));

    // register memory modules
    int count = 0;
    const goff_t USABLE_MEM  = (static_cast<goff_t>(2048) + 512) * 1024 * 1024;
    MainMemory &mem = MainMemory::get();
    for(size_t i = 0; i < pe_count; ++i) {
        if(pes[i].type() == m3::PEType::MEM) {
            // the first memory module hosts the FS image and other stuff
            if(count == 0) {
                mem.add(new MemoryModule(false, i, 0, USABLE_MEM));
                mem.add(new MemoryModule(true, i, USABLE_MEM, pes[i].mem_size() - USABLE_MEM));
            }
            else
                mem.add(new MemoryModule(true, i, 0, pes[i].mem_size()));
            count++;
        }
        else
            last_pe_id = i;
    }
}

peid_t Platform::kernel_pe() {
    // gem5 initializes the peid for us
    return m3::env()->pe;
}
peid_t Platform::first_pe() {
    return m3::env()->pe + 1;
}
peid_t Platform::last_pe() {
    return last_pe_id;
}

goff_t Platform::def_recvbuf(peid_t no) {
    if(pe(no).has_virtmem())
        return RECVBUF_SPACE;
    return pe(no).mem_size() - RECVBUF_SIZE_SPM;
}

}

File: src/apps/rust/kernel/src/arch/gem5/platform.rs
use base::cell::StaticCell;
use base::cfg;
use base::envdata;
use base::GlobAddr;
use base::goff;
use base::kif::{PEDesc, PEType};
use base::dtu::PEId;

use arch::kdtu::KDTU;
use mem;
use pes::VPEDesc;
use platform;

const USABLE_MEM: usize     = (1024 + 256) * 1024 * 1024;

static LAST_PE: StaticCell<PEId> = StaticCell::new(0);

pub fn init() -> platform::KEnv {
    // read kernel env
    let kenv_addr = GlobAddr::new(envdata::get().kenv);
    let kenv: platform::KEnv = KDTU::get().read_obj(
        &VPEDesc::new_mem(kenv_addr.pe()), kenv_addr.offset()
    );

    let mut count = 0;
    let mem: &mut mem::MainMemory = mem::get();
    for i in 0..kenv.pe_count as usize {
        let pedesc = PEDesc::new_from(kenv.pes[i]);
        if pedesc.pe_type() == PEType::MEM {
            let mut m = mem::MemMod::new(i, 0, pedesc.mem_size());
            if count == 0 {
                // allocate FS image
                m.allocate(USABLE_MEM, 1).expect("Not enough space for FS image");
            }
            mem.add(m);
            count += 1;
        }
        else {
            LAST_PE.set(i);
        }
    }

    kenv
}

pub fn kernel_pe() -> PEId {
    envdata::get().pe_id as PEId
}
pub fn first_user_pe() -> PEId {
    kernel_pe() + 1
}
pub fn last_user_pe() -> PEId {
    *LAST_PE
}

pub fn default_rcvbuf(pe: PEId) -> goff {
    if platform::pe_desc(pe).has_virtmem() {
        return cfg::RECVBUF_SPACE as goff
    }
    else {
        (platform::pe_desc(pe).mem_size() - cfg::RECVBUF_SIZE_SPM) as goff
    }
}
pub fn rcvbufs_size(pe: PEId) -> usize {
    match platform::pe_desc(pe).has_virtmem() {
        true    => cfg::RECVBUF_SIZE,
        false   => cfg::RECVBUF_SIZE_SPM,
    }
}
